File: packages/tiny-mcp-client-rust/src/http-transport.js
import { createRequire } from "node:module";
import { PassThrough } from "node:stream";
import { readLines } from "./layer.js";
import { OAuthMetadataDiscovery, parseBearerWwwAuthenticateHeader } from "./oauth-discovery.js";
import { createOAuthClientProvider } from "./oauth/provider.js";
import { OAuthError } from "./oauth/tokens.js";
import { fetchMcpResponse, readBoundedResponseText } from "./oauth/http.js";
const { NativeHttpTransport, NativeSseParser, httpResponseKind, validateRequestTimeout } = createRequire(import.meta.url)("./tiny-mcp-client-rust.node");

export class HttpTransportError extends Error {
  constructor(message, status, method, rpcMethod) {
    super(message);
    this.name = "HttpTransportError";
    this.status = status;
    this.method = method;
    this.rpcMethod = rpcMethod;
  }
}

export class HttpTransport {
  #state;
  #url;
  #mode;
  #endpoint;
  #endpointReady;
  #resolveEndpoint;
  #rejectEndpoint;
  #headers;
  #fetch;
  #warning;
  #provider;
  #discovery;
  #controllers = new Map();
  #fetches = new Set();
  #oauthControllers = new Set();
  #readers = new Set();
  #initializing = false;
  #resolveClosed;
  #resolveCloseReason;
  #read = new PassThrough();
  #write = new PassThrough();
  constructor({ url, mode = "streamable-http", headers = {}, fetch, oauth, oauthDiscoveryCache, onWarning, maxResponseBytes = 16 * 1024 * 1024 }) {
    try { this.#state = new NativeHttpTransport(maxResponseBytes); }
    catch (error) { throw new Error(error.message); }
    this.#url = url;
    this.#mode = mode;
    this.#headers = headers;
    this.#fetch = fetch;
    this.#warning = onWarning;
    this.#provider = oauth === undefined ? undefined : createOAuthClientProvider(oauth);
    this.#discovery = oauth === undefined ? undefined : new OAuthMetadataDiscovery({ fetch: this.#fetchWithAbort.bind(this), cache: oauthDiscoveryCache });
    this.readable = this.#read;
    this.writable = this.#write;
    this.closeReason = new Promise(resolve => { this.#resolveCloseReason = resolve; });
    this.closed = new Promise(resolve => { this.#resolveClosed = resolve; });
    for (const stream of [this.#read, this.#write]) stream.once("error", error => {
      this.dispose(error instanceof Error ? error : new Error(String(error)));
    });
    void this.#consumeLines().catch(error => { this.dispose(error instanceof Error ? error : new Error(String(error))); });
  }
  filterTools(tools, reset = true) {
    if (reset) this.#state.clearTools();
    const accepted = [];
    for (const tool of tools) {
      try { this.#state.filterTool(tool.name, tool.inputSchema); accepted.push(tool); }
      catch (error) { this.#warning?.(`Rejected MCP tool ${tool.name}: ${error instanceof Error ? error.message : String(error)}`); }
    }
    return accepted;
  }
  async completeInitialization(options) {
    if (options.timeoutMs !== null) validateRequestTimeout(options.timeoutMs, "timeoutMs");
    if (options.protocolVersion !== undefined) this.#state.setLegacyVersion(options.protocolVersion);
    this.#maybeStartGet();
    const deadline = options.timeoutMs > 0 ? AbortSignal.timeout(Math.ceil(options.timeoutMs)) : undefined;
    const signals = [options.signal, deadline].filter(signal => signal !== undefined);
    const signal = signals.length === 0 ? new AbortController().signal : AbortSignal.any(signals);
    signal.throwIfAborted();
    const controller = new AbortController();
    const abort = () => controller.abort(signal.reason);
    signal.addEventListener("abort", abort, { once: true });
    const line = '{"jsonrpc":"2.0","method":"notifications/initialized"}';
    try {
      await this.#sendPost(line, this.#state.prepare(line), controller);
      signal.throwIfAborted();
      if (this.#state.disposed) throw await this.closeReason;
    } catch (error) {
      if (error instanceof HttpTransportError)
        throw new HttpTransportError(error.message, error.status, error.method, "notifications/initialized");
      throw error;
    } finally {
      signal.removeEventListener("abort", abort);
    }
  }
  dispose(reason = new Error("HTTP transport disposed")) {
    const disposal = this.#state.dispose();
    if (disposal === null) return;
    this.#resolveCloseReason?.(reason);
    this.#resolveCloseReason = undefined;
    this.#rejectEndpoint?.(reason);
    this.#resolveEndpoint = undefined;
    this.#rejectEndpoint = undefined;
    for (const slot of disposal.slots) this.#controllers.get(slot)?.abort(reason);
    this.#controllers.clear();
    for (const controller of this.#fetches) controller.abort(reason);
    this.#fetches.clear();
    for (const controller of this.#oauthControllers) controller.abort(reason);
    this.#oauthControllers.clear();
    for (const reader of this.#readers) void reader.cancel().catch(() => undefined);
    this.#readers.clear();
    for (const stream of [this.#read, this.#write]) {
      if (!stream.destroyed && !stream.writableEnded) stream.end();
    }
    void this.#close(disposal.session, reason);
  }
  async #close(session, reason) {
    let closeReason = reason;
    if (session !== null) {
      const controller = new AbortController();
      let timeout;
      const expired = new Promise((_resolve, reject) => {
        timeout = setTimeout(() => {
          const error = new Error("HTTP transport session termination timed out");
          controller.abort(error);
          reject(error);
        }, 1000);
      });
      try { await Promise.race([this.#terminate(session, controller.signal), expired]); }
      catch (error) { closeReason = error instanceof Error ? error : new Error(String(error)); }
      finally { clearTimeout(timeout); }
    }
    const resolve = this.#resolveClosed;
    this.#resolveClosed = undefined;
    resolve?.({ reason: closeReason });
  }
  async #fetchWithAbort(input, init = {}, controller = new AbortController()) {
    if (this.#state.disposed) throw new Error("HTTP transport disposed");
    controller.signal.throwIfAborted();
    const signal = init.signal == null ? controller.signal : AbortSignal.any([init.signal, controller.signal]);
    signal.throwIfAborted();
    this.#fetches.add(controller);
    try { return await fetchMcpResponse(this.#fetch ?? globalThis.fetch, input, { ...init, signal }); }
    finally { this.#fetches.delete(controller); }
  }
  async #consumeLines() {
    for await (const line of readLines(this.#write)) {
      if (this.#state.disposed || line.length === 0) continue;
      const post = this.#state.prepare(line);
      if (post.cancelled) { this.#controllers.get(post.cancelSlot)?.abort(); continue; }
      const work = this.#sendPost(line, post);
      if (post.ordered) {
        try { await work; }
        catch (error) { this.dispose(error instanceof Error ? error : new Error(String(error))); }
      }
      else void work.catch(error => { this.dispose(error instanceof Error ? error : new Error(String(error))); });
    }
  }
  async #sendPost(line, post, completionController) {
    if (post.initializing) this.#initializing = true;
    const slot = post.slot;
    const controller = completionController ?? (slot === null ? undefined : new AbortController());
    if (controller !== undefined) this.#controllers.set(slot, controller);
    try {
      const endpoint = this.#mode === "sse" ? await this.#ensureEndpoint() : this.#url;
      if (this.#state.disposed) return;
      const response = await this.#fetchRetry("POST", post, line, controller, endpoint);
      if (this.#state.disposed || controller?.signal.aborted) { void response.body?.cancel().catch(() => undefined); return; }
      if (post.hasSession && response.status === 404) {
        void response.body?.cancel().catch(() => undefined);
        this.#state.expireSession();
        this.dispose(new Error("HTTP transport session expired (404 response)"));
        return;
      }
      if (response.status >= 400) {
        const body = (await readBoundedResponseText(response, this.#state.maxResponseBytes, this.#readers, controller?.signal)).trim();
        const errorLine = post.errorLine(response.status, body);
        if (errorLine !== null) { this.#emit(errorLine); return; }
        throw this.#failure("POST", response, body);
      }
      if (this.#state.disposed || controller?.signal.aborted) { void response.body?.cancel().catch(() => undefined); return; }
      if (this.#mode !== "sse" && !post.modern) { this.#state.captureSession(response.headers.get("Mcp-Session-Id")); if (!post.ordered) this.#maybeStartGet(); }
      if (controller !== undefined) await this.#forward(response, controller.signal, post.responseContext());
      else if (post.ordered) { await this.#forward(response, undefined, undefined, post.initializing); this.#maybeStartGet(); }
      else void this.#forward(response).catch(error => { this.dispose(error instanceof Error ? error : new Error(String(error))); });
    } catch (error) { if (!controller?.signal.aborted) throw error; }
    finally { this.#state.finish(post); if (controller !== undefined) this.#controllers.delete(slot); }
  }
  async #requestHeaders(method, post, session, signal) {
    const headers = new Headers(this.#headers);
    if (method === "POST") {
      let changes;
      try { changes = this.#state.postHeaders(post); }
      catch (error) { throw new Error(error.message); }
      for (const [name, value] of changes) {
        if (value === null) headers.delete(name);
        else headers.set(name, value);
      }
    } else if (method === "GET") {
      for (const [name, value] of this.#state.getHeaders()) headers.set(name, value);
    } else {
      headers.set("Mcp-Session-Id", session);
      headers.set("MCP-Protocol-Version", this.#state.legacyVersion);
    }
    signal?.throwIfAborted();
    const tokens = await this.#provider?.authorizeRequest?.({
      requestUrl: new URL(this.#url), headers, signal,
      fetch: (url, init) => fetchMcpResponse(this.#fetch ?? globalThis.fetch, url, {
        ...init, signal: signal === undefined ? init?.signal : init?.signal == null ? signal : AbortSignal.any([signal, init.signal]),
      }),
    });
    signal?.throwIfAborted();
    return { headers, tokens: tokens === undefined ? null : { ...tokens } };
  }
  #ensureEndpoint() {
    if (this.#endpointReady !== undefined) return this.#endpointReady;
    this.#endpointReady = new Promise((resolve, reject) => {
      this.#resolveEndpoint = resolve;
      this.#rejectEndpoint = reject;
    });
    void this.#consumeGet().catch(error => {
      this.dispose(error instanceof Error ? error : new Error(String(error)));
    });
    return this.#endpointReady;
  }
  #maybeStartGet() {
    if (!this.#state.beginGet()) return;
    void this.#consumeGet().catch(error => {
      if (!this.#state.disposed) this.dispose(error instanceof Error ? error : new Error(String(error)));
    });
  }
  async #consumeGet() {
    const response = await this.#fetchRetry("GET");
    if (this.#state.disposed) { void response.body?.cancel().catch(() => undefined); return; }
    if (response.status === 405) {
      void response.body?.cancel().catch(() => undefined);
      if (this.#mode === "sse") throw this.#failure("GET", response, "");
      return;
    }
    if (response.status === 404) {
      void response.body?.cancel().catch(() => undefined);
      this.#state.expireSession();
      throw new HttpTransportError("HTTP transport session expired (GET 404 response)", 404, "GET");
    }
    if (!response.ok) {
      const body = (await readBoundedResponseText(response, this.#state.maxResponseBytes, this.#readers)).trim();
      throw this.#failure("GET", response, body);
    }
    if (httpResponseKind(200, response.headers.get("Content-Type")) !== "sse") {
      void response.body?.cancel().catch(() => undefined);
      if (this.#mode === "sse") throw new Error("Legacy SSE GET returned an unsupported content type");
      return;
    }
    await this.#consumeSse(response, undefined, undefined, this.#mode === "sse");
    if (this.#mode === "sse") {
      if (!this.#state.disposed) throw new Error("Legacy SSE stream ended");
      return;
    }
    if (this.#state.finishGet()) this.#maybeStartGet();
  }
  async #terminate(session, signal) {
    const { headers } = await this.#requestHeaders("DELETE", undefined, session, signal);
    signal.throwIfAborted();
    const response = await fetchMcpResponse(this.#fetch ?? globalThis.fetch, this.#url, { method: "DELETE", headers, signal });
    if (signal.aborted) { void response.body?.cancel().catch(() => undefined); signal.throwIfAborted(); }
    if (response.ok || response.status === 405) { void response.body?.cancel().catch(() => undefined); return; }
    const body = (await readBoundedResponseText(response, this.#state.maxResponseBytes, this.#readers, signal)).trim();
    throw this.#failure("DELETE", response, body);
  }
  #failure(method, response, body) {
    const status = `${response.status} ${response.statusText}`.trim();
    return new HttpTransportError(`HTTP transport ${method} failed (${status})${body.length === 0 ? "" : `: ${body}`}`, response.status, method);
  }
  async #fetchRetry(method, post, body, controller = new AbortController(), endpoint = this.#url) {
    this.#oauthControllers.add(controller);
    try {
      let response;
      for (let attempt = 0; attempt < 2; attempt++) {
        controller.signal.throwIfAborted();
        const { headers, tokens } = await this.#requestHeaders(method, post, undefined, controller.signal);
        const requestHeaders = new Headers(headers);
        controller.signal.throwIfAborted();
        response = await this.#fetchWithAbort(endpoint, { method, headers, body }, controller);
        if (attempt === 0 && await this.#unauthorized(response, controller.signal, requestHeaders, tokens)) continue;
        break;
      }
      if (this.#provider !== undefined && (response.status === 401 || response.status === 403)) {
        const challenge = parseBearerWwwAuthenticateHeader(response.headers.get("WWW-Authenticate"));
        const error = challenge?.params.error;
        if (error !== undefined && error.length > 0) {
          void response.body?.cancel().catch(() => undefined);
          throw new OAuthError({ error, error_description: challenge.params.error_description, error_uri: challenge.params.error_uri }, response.status);
        }
      }
      return response;
    } finally { this.#oauthControllers.delete(controller); }
  }
  async #unauthorized(response, signal, requestHeaders, presentedTokens) {
    if (response.status !== 401 || this.#provider === undefined || this.#discovery === undefined) return false;
    const challenge = parseBearerWwwAuthenticateHeader(response.headers.get("WWW-Authenticate"));
    try {
      const discovery = await this.#discovery.discover(this.#url, { resourceMetadataUrl: challenge?.params.resource_metadata, signal });
      const providerResponse = response.clone();
      let result;
      try {
        result = await this.#provider.handleUnauthorized({
          requestUrl: new URL(this.#url), response: providerResponse, challenge, discovery,
          signal, requestHeaders: new Headers(requestHeaders), presentedTokens,
          fetch: (url, init) => fetchMcpResponse(this.#fetch ?? globalThis.fetch, url, {
            ...init, signal: init?.signal == null ? signal : AbortSignal.any([signal, init.signal]),
          }),
        });
        signal.throwIfAborted();
      } finally { void providerResponse.body?.cancel().catch(() => undefined); }
      if (result.action === "retry") { void response.body?.cancel().catch(() => undefined); return true; }
      if (result.error !== undefined) throw result.error;
      return false;
    } catch (error) { void response.body?.cancel().catch(() => undefined); throw error; }
  }
  async #forward(response, signal, context, stopAfterInitialization = false) {
    switch (httpResponseKind(response.status, response.headers.get("Content-Type"))) {
      case "ignore": void response.body?.cancel().catch(() => undefined); return;
      case "sse": await this.#consumeSse(response, signal, context, false, stopAfterInitialization); return;
      case "json": {
        const payload = await readBoundedResponseText(response, this.#state.maxResponseBytes, this.#readers, signal);
        if (payload.length === 0) { if (context !== undefined) throw new Error("MCP HTTP response body is empty"); return; }
        this.#emit(context === undefined ? JSON.stringify(JSON.parse(payload)) : context.validate(payload, false));
        return;
      }
      default: void response.body?.cancel().catch(() => undefined); throw new Error("HTTP transport POST returned an unsupported response content type");
    }
  }
  async #consumeSse(response, signal, context, acceptEndpoint = false, stopAfterInitialization = false) {
    if (response.body === null) return;
    const parser = new NativeSseParser(this.#state.maxResponseBytes, acceptEndpoint);
    const decoder = new TextDecoder("utf-8", { fatal: true });
    const reader = response.body.getReader();
    this.#readers.add(reader);
    const abort = () => { void reader.cancel().catch(() => undefined); };
    signal?.addEventListener("abort", abort, { once: true });
    if (signal?.aborted) abort();
    try {
      while (true) {
        const { done, value } = await reader.read();
        signal?.throwIfAborted();
        if (done) break;
        if (value === undefined) continue;
        this.#emitEvents(parser.push(decoder.decode(value, { stream: true })), context);
        if (context?.completed || (stopAfterInitialization && !this.#initializing)) { void reader.cancel().catch(() => undefined); return; }
        this.#state.setEventId(parser.lastEventId);
      }
      const trailing = decoder.decode();
      if (trailing.length > 0) { this.#emitEvents(parser.push(trailing), context); this.#state.setEventId(parser.lastEventId); }
      this.#emitEvents(parser.flush(), context);
      this.#state.setEventId(parser.lastEventId);
      if (context !== undefined && !context.completed) throw new Error("MCP HTTP stream ended before its final response");
    } catch (error) { void reader.cancel().catch(() => undefined); throw error; }
    finally { signal?.removeEventListener("abort", abort); this.#readers.delete(reader); reader.releaseLock(); }
  }
  #emitEvents(messages, context) {
    for (const message of messages) {
      if (message.event === "endpoint") {
        const endpoint = new URL(message.data, this.#url);
        const resource = new URL(this.#url);
        if (!["http:", "https:"].includes(endpoint.protocol) || endpoint.origin !== resource.origin
          || endpoint.username || endpoint.password || endpoint.hash) throw new Error("Unsafe legacy SSE endpoint");
        if (this.#endpoint !== undefined && this.#endpoint !== endpoint.href)
          throw new Error("Legacy SSE endpoint changed during the active connection");
        this.#endpoint = endpoint.href;
        this.#resolveEndpoint?.(endpoint.href);
        this.#resolveEndpoint = undefined;
        this.#rejectEndpoint = undefined;
        continue;
      }
      this.#emit(context === undefined ? message.data : context.validate(message.data, true));
      if (context?.completed) return;
    }
  }
  #emit(line) {
    if (this.#initializing && this.#state.captureInitialization(line)) { this.#initializing = false; this.#maybeStartGet(); }
    if (!this.#state.disposed && !this.#read.destroyed && !this.#read.writableEnded) this.#read.write(`${line}\n`);
  }
}

File: packages/tiny-mcp-client-rust/src/http_transport.rs
//! HTTP transport policy and ownership identities. The host executes HTTP/stream I/O.
use crate::{
    http::HttpResponseMessages,
    messages::{ParsedMessage, parse_message},
};
use mcp_protocol_rust::{
    json::{self, Value},
    strings::trim_ecmascript,
};
use std::collections::HashMap;
use std::sync::Arc;
use tiny_stdio_mcp_server_rust::headers::{self, ParameterHeader};

#[derive(Hash, PartialEq, Eq)]
enum Id {
    String(Vec<u16>),
    Number(u64),
}
impl Id {
    fn from_value(value: &Value) -> Option<Self> {
        match value {
            Value::String(value) => Some(Self::String(value.clone())),
            Value::Number(value) => Some(Self::Number(if *value == 0.0 {
                0
            } else if value.is_nan() {
                f64::NAN.to_bits()
            } else {
                value.to_bits()
            })),
            _ => None,
        }
    }
}
fn is_text(value: Option<&Value>, expected: &str) -> bool {
    matches!(value, Some(Value::String(value)) if value.iter().copied().eq(expected.encode_utf16()))
}
fn record(value: Option<&Value>) -> Option<&Value> {
    value.filter(|value| matches!(value, Value::Object(_)))
}
fn text(value: &str) -> Vec<u16> {
    value.encode_utf16().collect()
}
pub struct Post {
    pub message: Option<Arc<Value>>,
    pub modern: bool,
    pub ordered: bool,
    pub slot: Option<u64>,
    pub cancelled: bool,
    pub cancel_slot: Option<u64>,
    pub has_session: bool,
}
pub type HeaderChanges = Vec<(String, Option<Vec<u16>>)>;
impl Post {
    pub fn response_context(&self) -> Option<HttpResponseMessages> {
        self.message
            .as_ref()
            .filter(|message| message.get("id").is_some())
            .map(|message| {
                HttpResponseMessages::from_shared(message.clone())
                    .expect("parsed originating request")
            })
    }
    pub fn error_line(&self, status: u16, body: &[u16]) -> Option<Vec<u16>> {
        if !self.modern || status < 400 {
            return None;
        }
        let request = self
            .message
            .as_ref()
            .filter(|message| message.get("id").is_some())?;
        if !body.is_empty() {
            let mut context = self.response_context()?;
            if let Ok(line) = context.validate(body, false)
                && let ParsedMessage::Response(mut response) = parse_message(&line)
                && response.get("error").is_some()
            {
                super::http::order_properties(&mut response);
                return Some(text(&json::stringify(&response)));
            }
        }
        if status < 500 && is_text(request.get("method"), "server/discover") {
            return Some(text(&json::stringify(&Value::Object(vec![
                (text("jsonrpc"), Value::String(text("2.0"))),
                (text("id"), request.get("id").expect("request ID").clone()),
                (
                    text("error"),
                    Value::Object(vec![
                        (text("code"), Value::Number(-32601.0)),
                        (
                            text("message"),
                            Value::String(text("Modern discovery unavailable")),
                        ),
                    ]),
                ),
            ]))));
        }
        None
    }
}
#[derive(Debug, PartialEq)]
pub enum ResponseKind {
    Ignore,
    Json,
    Sse,
    Unsupported,
}
pub fn response_kind(status: u16, content_type: Option<&str>) -> ResponseKind {
    if status == 202 {
        return ResponseKind::Ignore;
    }
    let Some(content_type) = content_type else {
        return ResponseKind::Ignore;
    };
    let units = content_type
        .split(';')
        .next()
        .unwrap_or_default()
        .encode_utf16()
        .collect::<Vec<_>>();
    let units = trim_ecmascript(&units);
    let matches = |expected: &str| {
        units.len() == expected.len()
            && units
                .iter()
                .copied()
                .zip(expected.bytes())
                .all(|(left, right)| {
                    left == u16::from(right)
                        || (right.is_ascii_lowercase() && left == u16::from(right - 32))
                })
    };
    if matches("application/json") {
        ResponseKind::Json
    } else if matches("text/event-stream") {
        ResponseKind::Sse
    } else {
        ResponseKind::Unsupported
    }
}
pub struct Disposal {
    pub session: Option<Vec<u16>>,
    pub slots: Vec<u64>,
}
#[derive(Default)]
pub struct HttpState {
    session: Option<Vec<u16>>,
    event_id: Option<Vec<u16>>,
    get_started: bool,
    modern: bool,
    disposed: bool,
    next: u64,
    active: HashMap<Id, u64>,
    tool_headers: HashMap<Vec<u16>, Vec<ParameterHeader>>,
}
impl HttpState {
    pub fn disposed(&self) -> bool {
        self.disposed
    }
    pub fn active_count(&self) -> usize {
        self.active.len()
    }
    pub fn prepare(&mut self, line: &[u16]) -> Result<Post, &'static str> {
        if self.disposed {
            return Err("HTTP transport disposed");
        }
        let parsed = parse_message(line);
        let ordered = match &parsed {
            ParsedMessage::Request(message) => is_text(message.get("method"), "initialize"),
            ParsedMessage::Notification(message) => {
                is_text(message.get("method"), "notifications/initialized")
            }
            _ => false,
        };
        let notification = matches!(parsed, ParsedMessage::Notification(_));
        let message = match parsed {
            ParsedMessage::Request(message) | ParsedMessage::Notification(message) => {
                Some(Arc::new(message))
            }
            _ => None,
        };
        let modern = message
            .as_ref()
            .and_then(|message| record(message.get("params")))
            .and_then(|params| record(params.get("_meta")))
            .is_some_and(|metadata| {
                matches!(
                    metadata.get("io.modelcontextprotocol/protocolVersion"),
                    Some(Value::String(_))
                )
            });
        if modern {
            self.modern = true;
        }
        if ordered && !notification {
            self.modern = false;
        }
        let cancelled = self.modern
            && notification
            && message
                .as_ref()
                .is_some_and(|message| is_text(message.get("method"), "notifications/cancelled"));
        let cancel_slot = if cancelled {
            message
                .as_ref()
                .and_then(|message| record(message.get("params")))
                .and_then(|params| params.get("requestId"))
                .and_then(Id::from_value)
                .and_then(|id| self.active.get(&id).copied())
        } else {
            None
        };
        let slot = if !cancelled
            && modern
            && !notification
            && let Some(id) = message
                .as_ref()
                .and_then(|message| message.get("id"))
                .and_then(Id::from_value)
        {
            self.next = self
                .next
                .checked_add(1)
                .filter(|slot| *slot <= 9_007_199_254_740_991)
                .ok_or("HTTP transport request identity exhausted")?;
            self.active.insert(id, self.next);
            Some(self.next)
        } else {
            None
        };
        Ok(Post {
            message,
            modern,
            ordered,
            slot,
            cancelled,
            cancel_slot,
            has_session: !modern && self.session.is_some(),
        })
    }
    pub fn finish(&mut self, post: &Post) {
        if let Some(id) = post
            .message
            .as_ref()
            .and_then(|message| message.get("id"))
            .and_then(Id::from_value)
            && self.active.get(&id) == post.slot.as_ref()
        {
            self.active.remove(&id);
        }
    }
    pub fn clear_tools(&mut self) {
        self.tool_headers.clear();
    }
    pub fn filter_tool(&mut self, name: Vec<u16>, schema: &Value) -> Result<(), String> {
        match headers::get_parameter_headers(schema) {
            Ok(definitions) => {
                self.tool_headers.insert(name, definitions);
                Ok(())
            }
            Err(error) => {
                self.tool_headers.remove(&name);
                Err(error)
            }
        }
    }
    pub fn reject_tool(&mut self, name: &[u16]) {
        self.tool_headers.remove(name);
    }
    pub fn post_headers(&self, post: &Post) -> Result<HeaderChanges, String> {
        let mut changes = vec![
            (
                "Accept".into(),
                Some(text("application/json, text/event-stream")),
            ),
            ("Content-Type".into(), Some(text("application/json"))),
        ];
        let message = post.message.as_ref();
        if post.modern
            && let Some(message) = message
            && let Some(params) = record(message.get("params"))
            && let Some(metadata) = record(params.get("_meta"))
        {
            changes.push(("Mcp-Session-Id".into(), None));
            changes.push(("Last-Event-ID".into(), None));
            changes.push((
                "MCP-Protocol-Version".into(),
                metadata
                    .get("io.modelcontextprotocol/protocolVersion")
                    .and_then(|value| match value {
                        Value::String(value) => Some(value.clone()),
                        _ => None,
                    }),
            ));
            if message.get("id").is_some() {
                if let Some(Value::String(method)) = message.get("method") {
                    changes.push(("Mcp-Method".into(), Some(method.clone())));
                }
                let method = message.get("method");
                let name = params.get(if is_text(method, "resources/read") {
                    "uri"
                } else {
                    "name"
                });
                let named_method = ["tools/call", "resources/read", "prompts/get"]
                    .iter()
                    .any(|expected| is_text(method, expected));
                changes.push((
                    "Mcp-Name".into(),
                    if named_method && let Some(Value::String(name)) = name {
                        Some(headers::encode_value(name)?)
                    } else {
                        None
                    },
                ));
                if is_text(method, "tools/call")
                    && let Some(Value::String(name)) = params.get("name")
                    && let Some(definitions) = self.tool_headers.get(name)
                {
                    for definition in definitions {
                        changes.push((definition.name.clone(), None));
                    }
                    for (name, value) in headers::create_parameter_headers(
                        definitions,
                        params.get("arguments").unwrap_or(&Value::Null),
                    )? {
                        changes.push((name, Some(value)));
                    }
                }
            }
        } else if let Some(session) = &self.session {
            changes.push(("Mcp-Session-Id".into(), Some(session.clone())));
            changes.push(("MCP-Protocol-Version".into(), Some(text("2025-03-26"))));
        }
        Ok(changes)
    }
    pub fn get_headers(&self) -> Vec<(String, Vec<u16>)> {
        let mut headers = vec![("Accept".into(), text("text/event-stream"))];
        if let Some(session) = &self.session {
            headers.push(("Mcp-Session-Id".into(), session.clone()));
            headers.push(("MCP-Protocol-Version".into(), text("2025-03-26")));
        }
        if let Some(cursor) = &self.event_id {
            headers.push(("Last-Event-ID".into(), cursor.clone()));
        }
        headers
    }
    pub fn capture_session(&mut self, session: Option<&[u16]>) -> Result<(), &'static str> {
        if let Some(session) = session.filter(|value| !value.is_empty()) {
            if self
                .session
                .as_ref()
                .is_some_and(|current| current != session)
            {
                return Err("HTTP transport response changed active session ID");
            }
            self.session = Some(session.to_vec());
        }
        Ok(())
    }
    pub fn expire_session(&mut self) {
        self.session = None;
    }
    pub fn begin_get(&mut self) -> bool {
        if self.disposed || self.session.is_none() || self.get_started {
            return false;
        }
        self.get_started = true;
        true
    }
    pub fn set_event_id(&mut self, cursor: Option<Vec<u16>>) {
        self.event_id = cursor;
    }
    pub fn finish_get(&mut self) -> bool {
        self.get_started = false;
        !self.disposed && self.session.is_some() && self.event_id.is_some()
    }
    pub fn dispose(&mut self) -> Option<Disposal> {
        if self.disposed {
            return None;
        }
        self.disposed = true;
        self.tool_headers.clear();
        let slots = self.active.drain().map(|(_, slot)| slot).collect();
        Some(Disposal {
            session: self.session.take(),
            slots,
        })
    }
}
